File: Python/06_union_and_intersection_of_2_sorted_arrays.py
"""
Idea:
Union: if element is smaller, print and go to next index, if both elements are same, print only one, and go to
next element of both array. At the end, go through both array's remaining elements( if any ).
Intersection: if element is smaller, go to next index, if both elements are same, print only one, and go to
next element of both array.
"""


def findUnion(arr1, arr2, len1, len2):
    i = j = 0
    ans = []
    while i < len1 and j < len2:
        # checking if elements are equal or smaller to change indices
        if arr1[i] < arr2[j]:
            ans.append(arr1[i])
            i += 1
        elif arr1[i] > arr2[j]:
            ans.append(arr2[j])
            j += 1
        else:
            # appending only 1 element, when both elements are same
            ans.append(arr1[i])
            i += 1
            j += 1

    # appending remaining elements
    while i < len1:
        ans.append(arr1[i])
        i += 1
    while j < len2:
        ans.append(arr2[j])
        j += 1
    print(ans)


def findIntersection(arr1, arr2, len1, len2):
    i = j = 0
    ans = []
    while i < len1 and j < len2:
        # checking if elements are equal or smaller to change indices
        if arr1[i] < arr2[j]:
            i += 1
        elif arr1[i] > arr2[j]:
            j += 1
        else:
            # appending only 1 element, when both elements are same
            ans.append(arr1[i])
            i += 1
            j += 1

    print(ans)


if __name__ == '__main__':
    arr1 = [1, 2, 4, 5, 6]
    arr2 = [2, 3, 5, 7]
    findUnion(arr1, arr2, len(arr1), len(arr2))
    findIntersection(arr1, arr2, len(arr1), len(arr2))

File: C++/Array/06_union_and_intersection_of_2_sorted_arrays.cpp
#include <bits/stdc++.h>
using namespace std;

/*
Idea:
Union: if element is smaller, print and go to next index, if both elements are same, print only one, and go to
next element of both array. At the end, go through both array's remaining elements( if any ).
Intersection: if element is smaller, go to next index, if both elements are same, print only one, and go to
next element of both array.
*/
void findUnion(int* a, int* b, int m, int n)
{
    int i = 0, j = 0;
    while (i < m && j < n)
    {
        //checking if elements are equal or smaller to change indices
        if (a[i] < b[j])
            cout << a[i++] << " ";
        else if (a[i] > b[j])
            cout << b[j++] << " ";
        else
        {
            //printing only 1 element, when both elements are same
            cout << a[i++] << " ";
            ++j;
        }
    }
    //printing remaining elements
    while (i < m)
        cout << a[i++] << " ";
    while (j < n)
        cout << b[j++] << " ";
    cout << "\n";
}

void findIntersection(int* a, int* b, int m, int n)
{
    int i = 0, j = 0;
    while (i < m && j < n)
    {
        //checking if elements are equal or smaller to change indices
        if (a[i] < b[j])
            ++i;
        else if (a[i] > b[j])
            ++j;
        else
        {
            //printing only 1 element, when both elements are same
            cout << a[i++] << " ";
            ++j;
        }
    }
    cout << "\n";
}

int main()
{
    int arr1[] = { 1, 2, 4, 5, 6 };
    int arr2[] = { 2, 3, 5, 7 };

    int m = sizeof(arr1) / sizeof(arr1[0]);
    int n = sizeof(arr2) / sizeof(arr2[0]);

    findIntersection(arr1, arr2, m, n);
    findUnion(arr1, arr2, m, n);

    return 0;
}